pkg/drift/driftignore.go

```diff
@@ -0,0 +1,143 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package drift
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/abcxyz/pkg/logging"
+
+	"github.com/abcxyz/guardian/pkg/drift/assets"
+	"github.com/abcxyz/guardian/pkg/drift/iam"
+)
+
+type ignoredAssets struct {
+	iamAssets map[string]struct{}
+	projects  map[string]struct{}
+	folders   map[string]struct{}
+}
+
+// ignoredProjectPattern is a Regex pattern used to identify projects that should be ignored.
+var ignoredProjectPattern = regexp.MustCompile(`^\/organizations\/(?:\d*)\/projects\/(\d*)$`)
+
+// ignoredFolderPattern is a  Regex pattern used to identify folders that should be ignored.
+var ignoredFolderPattern = regexp.MustCompile(`^\/organizations\/(?:\d*)\/folders\/(\d*)$`)
+
+// filterIgnored removes any asset iam that is in the ignored assets.
+func filterIgnored(values map[string]*iam.AssetIAM, ignored *ignoredAssets) map[string]*iam.AssetIAM {
+	filtered := make(map[string]*iam.AssetIAM)
+	for k, a := range values {
+		if a.ResourceType == assets.Project {
+			if _, ok := ignored.projects[a.ResourceID]; !ok {
+				filtered[k] = a
+			}
+		} else if a.ResourceType == assets.Folder {
+			if _, ok := ignored.folders[a.ResourceID]; !ok {
+				filtered[k] = a
+			}
+		} else { // Handle default case so we do not accidentally drop values.
+			filtered[k] = a
+		}
+	}
+
+	return filtered
+}
+
+// expandGraph traverses the asset hierarchy graph and adds any nested folders or projects beneath every ignored asset.
+func expandGraph(ignored *ignoredAssets, hierarchyGraph *assets.HierarchyGraph) (*ignoredAssets, error) {
+	ignoredProjects := ignored.projects
+	ignoredFolders := ignored.folders
+
+	// Traverse the hierarchy
+	for folderID := range ignored.folders {
+		ids, err := assets.FoldersBeneath(folderID, hierarchyGraph)
+		if err != nil {
+			return nil, fmt.Errorf("failed to traverse hierarchy for folder with ID %s: %w", folderID, err)
+		}
+		ignoredFolders = mergeSets(ignoredFolders, ids)
+	}
+	for folderID := range ignoredFolders {
+		ignoredProjects = addListToSet(ignoredProjects, hierarchyGraph.IDToNodes[folderID].ProjectIDs)
+	}
+
+	return &ignoredAssets{
+		iamAssets: ignored.iamAssets,
+		projects:  ignoredProjects,
+		folders:   ignoredFolders,
+	}, nil
+}
+
+// driftignore parses the driftignore file into a set.
+// Go doesn't implement set so we use a map of boolean values all set to true.
+func driftignore(ctx context.Context, fname string) (*ignoredAssets, error) {
```

Comment:
Likely I'm missing some context, but how about using a YAML or JSON file here to avoid the complexity of writing a parser?

Reply:
I think this is the likely direction this will go - however we don't yet have enough use cases to understand what is needed quite yet. I'll add a TODO to explore doing this but I want to delay until we have 1-2 real customers using it and we get real feedback (I will be the first customer).

I'll leave as is for now until we collect more feedback though. Let me know if you think this is appropriate?

Reply:
FIne by me, as long as you've considered it.

Reply:
https://github.com/abcxyz/guardian/issues/105